example/src/CrudWebserverDb.hs

```diff
@@ -0,0 +1,415 @@
+--
+------------------------------------------------------------------------
+
+module CrudWebserverDb where
+
+import           Control.Concurrent
+                   (threadDelay)
+import           Control.Concurrent.Async
+                   (async)
+import           Control.Monad.IO.Class
+                   (MonadIO, liftIO)
+import           Control.Monad.Logger
+                   (runNoLoggingT)
+import           Control.Monad.Reader
+                   (ReaderT, ask, runReaderT)
+import           Data.Aeson
+                   (FromJSON, ToJSON, object, parseJSON, toJSON,
+                   withObject, (.:), (.=))
+import           Data.Char
+                   (isPrint)
+import           Data.Functor.Classes
+                   (Show1)
+import           Data.Proxy
+                   (Proxy(Proxy))
+import           Data.String.Conversions
+                   (cs)
+import           Data.Text
+                   (Text)
+import qualified Data.Text                 as T
+import           Database.Persist.Sqlite
+                   (ConnectionPool, Key, createSqlitePool, delete, get,
+                   insert, runMigration, runSqlPersistMPool,
+                   runSqlPool, update, (+=.))
+import           Database.Persist.TH
+                   (mkMigrate, mkPersist, persistLowerCase, share,
+                   sqlSettings)
+import           Network.HTTP.Client
+                   (Manager, defaultManagerSettings, newManager)
+import qualified Network.Wai.Handler.Warp  as Warp
+import           Servant
+                   ((:<|>)(..), (:>), Application, Capture, Get, JSON,
+                   Post, ReqBody, Server, serve)
+import           Servant.Client
+                   (BaseUrl(..), ClientEnv(..), ClientM, Scheme(Http),
+                   client, runClientM)
+import           Test.QuickCheck
+                   (Arbitrary, Property, arbitrary, elements,
+                   frequency, ioProperty, property, shrink, suchThat,
+                   (===))
+import           Test.QuickCheck.Instances
+                   ()
+
+import           Test.StateMachine
+
+------------------------------------------------------------------------
+
+-- Our webserver deals with users that have a name and an age. The
+-- following is the Persistent library's way of creating the datatype
+-- and corresponding database table.
+
+share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
+User
+  name Text
+  age  Int
+  deriving Eq Read Show
+|]
+
+-- The webserver will also need to (un)marshall the datatype into JSON.
+
+instance FromJSON User where
+  parseJSON = withObject "User" $ \ v ->
+    User <$> v .: "name"
+         <*> v .: "age"
+
+instance ToJSON User where
+  toJSON (User name age) =
+    object [ "name" .= name
+           , "age"  .= age
+           ]
+
+-- For testing we will also need to generate arbitrary users.
+
+instance Arbitrary User where
+  arbitrary = User <$> (T.pack <$> (arbitrary `suchThat` all isPrint))
```

Comment:
Should maybe have this be
```haskell
arbitrary = User <$> (T.pack <$> (listOf (arbitrary `suchThat` isPrint)))
                 <*> arbitrary
```
To be more likely to succeed in generating longer names.